Lib/functools.py

```diff
@@ -933,6 +933,7 @@ def __init__(self, func):
 
         self.dispatcher = singledispatch(func)
         self.func = func
+        self.attrname = None
```

Comment:
Do we actually need to set it to `None` here? In what situation would `__get__` be called without `__set_name__` having first been called to set `attrname`?

Reply:
This was copied from the original PR. I am not very familiar with descriptor semantics, but it might indeed not be required. I updated the code and `functools` tests pass locally so I will push it so the branch.

Reply:
According to https://peps.python.org/pep-0487/#implementation-details:

> `type.__new__` calls the `__set_name__` hooks on the descriptor after the new class has been initialized.

So in the examples in https://docs.python.org/3/howto/descriptor.html#customized-names, `__set_name__` should be called on the `name` and `age` descriptor instances immediately after the `Person` class has been created